storage/s3.go

```diff
@@ -126,13 +126,72 @@ func (s *S3) Get(ctx context.Context, path string) (io.ReadCloser, error) {
 		Bucket: &s.cfg.Bucket,
 		Key:    &path,
 	}
+
 	out, err := s.s3client.GetObject(ctx, input)
 	if err != nil {
 		return nil, errors.Wrapf(err, "fail to get S3 object %v", path)
 	}
 	return out.Body, nil
 }
 
+// GetWithRetries function downloads the object from S3 and writes it to the
+// writer. It returns the number of bytes written to the writer.
+//
+// It implements a retry mechanism to handle if connection is closed before
+// the end of the download.
+func (s *S3) GetWithRetries(ctx context.Context, path string, writer io.Writer) (int64, error) {
+	path = fullPath(path)
+	log := logger.Get(ctx)
+	log.WithField("path", path).Info("Get object with retries")
+
+	size, err := s.Size(ctx, path)
+	if err != nil {
+		return -1, errors.Wrapf(err, "get size of S3 object %v", path)
+	}
+
+	var (
+		offset  int64
+		attempt = 1
+	)
+
+	for {
+		// Log if we are resuming a download
+		if offset != 0 {
+			log.Infof("Downloading from offset %d...", offset)
```

Comment:
suggestion(non-blocking): add the offset as a logger field
```suggestion
			log.WithField("offset", offset).Infof("Downloading from offset %d...", offset)
```

Reply:
Ack